content/300-guides/100-prisma-guides/100-prisma-client-transactions-guide.mdx

```diff
@@ -0,0 +1,773 @@
+await prisma.team.update({
+  data:{
+    customerId: customer.id
+  },
+  where: {
+    id: teamId,
+  }
+})
+```
+
+This example has a problem: you can only run the logic _once_. Consider the following scenario:
+
+1. Stripe creates a new customer and subscription, and returns a customer ID
+2. Updating the team **fails** - the team is not marked as a customer in the Slack database
+3. The customer is charged by Stripe, but paid features are not unlocked in Slack because the team lacks a valid `customerId`
+4. Running the same code again either:
+
+   * Results in an error because the team (defined by `externalId`) already exists - Stripe never returns a customer ID
+   * If `externalId` is not subject to a unique constraint, Stripe creates a yet another subscription (**not idempotent**)
+
+You cannot re-run this code in case of error and you cannot change to another plan without being charged twice. 
+
+The following refactor (highlighted) introduces a mechanism that checks if a subscription already exists, and either creates the description or updates the existing subscription (which will remain unchanged if the input is identical):
+
+```ts highlight=12-27;normal
+// Calculate the number of users times the cost per user
+const numTeammates = await client.user.count({
+  where: {
+    teams: {
+      some: {
+        id: teamId,
+      },
+    },
+  },
+})
+
+// Find customer in Stripe
+let customer = await stripe.customers.get({ externalId: teamID })
+
+if (customer) {
+  customer = await stripe.customers.update({ // If team already exists, update
+    externalId: teamId,
+    plan: "plan_id",
+    quantity: numTeammates
+  })
+} else {
+  customer = await stripe.customers.create({ // If team does not exist, create customer
+    externalId: teamId,
+    plan: "plan_id",
+    quantity: numTeammates
+  })
+}
+
+// Update the team with the customer id to indicate that they are a customer
+// and support querying this customer in Stripe from our application code.
+await client.team.update({
+  data:{
+    customerId: customer.id
+  },
+  where: {
+    id: teamId,
+  }
+})
+```
+
+You can now retry the same logic multiple times with the same input without adverse effect. To further enhance this example, you can introduce a mechanism whereby the subscription is cancelled or temporarily deactivated if the update does not succeed after a set number of attempts.
+
+### Optimistic concurrency control
+
+Optimistic concurrency control (OCC) is a model for handling concurrent operations on a single entity that does not rely on 🔒 locking. Instead, we **optimistically** assume that a record will remain unchanged in between reading and writing, and use some kind of concurrency token (a timestamp or version field) to detect if a record has changed. 
+
+If a conflict ❌ occurs (someone else has changed the record since you read it), cancel the transaction. Depending on your scenario, you can then:
```

Comment:
```suggestion
If a conflict ❌ occurs (someone else has changed the record since you read it), you cancel the transaction. Depending on your scenario, you can then:
```